Index: PDFGraphics2D.java
===================================================================
--- PDFGraphics2D.java  (Revision 570024)
+++ PDFGraphics2D.java  (Arbeitskopie)
@@ -719,13 +719,15 @@
         }

         c = getColor();
-        if (graphicsState.setColor(c)) {
+        graphicsState.setColor(c);
+        // if (graphicsState.setColor(c)) {
             applyColor(c, false);
-        }
+        // }
         c = getBackground();
-        if (graphicsState.setBackColor(c)) {
+        graphicsState.setBackColor(c);
+        // if (graphicsState.setBackColor(c)) {
             applyColor(c, true);
-        }
+        // }

         Paint paint = getPaint();
         if (graphicsState.setPaint(paint)) {
@@ -1718,13 +1720,15 @@
         }

         c = getColor();
-        if (graphicsState.setColor(c)) {
+        graphicsState.setColor(c);
+        // if (graphicsState.setColor(c)) {
             applyColor(c, true);
-        }
+        // }
         c = getBackground();
-        if (graphicsState.setBackColor(c)) {
+        graphicsState.setBackColor(c);
+        // if (graphicsState.setBackColor(c)) {
             applyColor(c, false);
-        }
+        // }

         Paint paint = getPaint();
         if (graphicsState.setPaint(paint)) {
