Index: src/resources/skins/common/xslt/fo/document2fo.xsl
===================================================================
RCS file: /home/cvspublic/xml-forrest/src/resources/skins/common/xslt/fo/document2fo.xsl,v
retrieving revision 1.4
diff -u -r1.4 document2fo.xsl
--- src/resources/skins/common/xslt/fo/document2fo.xsl	4 Dec 2002 14:14:29 -0000	1.4
+++ src/resources/skins/common/xslt/fo/document2fo.xsl	5 Dec 2002 09:23:13 -0000
@@ -93,6 +93,25 @@
     </fo:root>
   </xsl:template>
 
+  <xsl:template name="info">
+    <fo:block-container font-style="italic" absolute-position="absolute"
+              left="0pt" top="0pt" right="6.25in" bottom="150pt"
+              font-size="10pt">
+      <fo:block text-align="center">
+        <fo:basic-link color="lightgrey"
+                       external-destination="http://xml.apache.org/fop/dev">
+          Created by: FOP 1.0dev
+        </fo:basic-link>
+      </fo:block>
+      <fo:block text-align="center">
+        <fo:basic-link color="lightgrey"
+                       external-destination="http://xml.apache.org/fop/dev">
+          http://xml.apache.org/fop/dev
+        </fo:basic-link>
+      </fo:block>
+    </fo:block-container>
+  </xsl:template>
+
   <xsl:template match="document">
     <fo:title><xsl:value-of select="header/title"/></fo:title>
     
@@ -107,6 +126,7 @@
         text-align="start">
         Page <fo:page-number/>
       </fo:block>
+      <xsl:call-template name="info"/>
     </fo:static-content>
 
     <fo:static-content flow-name="even-header">
@@ -115,6 +135,7 @@
         font-style="italic">
         <xsl:value-of select="header/title"/>
       </fo:block>
+      <xsl:call-template name="info"/>
     </fo:static-content>
 
     <fo:static-content flow-name="even-footer">
@@ -136,6 +157,7 @@
         font-style="italic">
         <xsl:value-of select="header/title"/>
       </fo:block>
+      <xsl:call-template name="info"/>
     </fo:static-content>
     
     <fo:static-content flow-name="odd-footer">
